source/common/ext_authz/ext_authz_impl.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include <chrono>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "envoy/ext_authz/ext_authz.h"
+#include "envoy/grpc/async_client.h"
+#include "envoy/grpc/async_client_manager.h"
+#include "envoy/http/filter.h"
+#include "envoy/http/header_map.h"
+#include "envoy/http/protocol.h"
+#include "envoy/network/address.h"
+#include "envoy/network/connection.h"
+#include "envoy/network/filter.h"
+#include "envoy/tracing/http_tracer.h"
+#include "envoy/upstream/cluster_manager.h"
+
+#include "common/singleton/const_singleton.h"
+
+namespace Envoy {
+namespace ExtAuthz {
+
+typedef Grpc::TypedAsyncRequestCallbacks<envoy::service::auth::v2::CheckResponse>
+    ExtAuthzAsyncCallbacks;
+
+struct ConstantValues {
+  const std::string TraceStatus = "ext_authz_status";
+  const std::string TraceUnauthz = "ext_authz_unauthorized";
+  const std::string TraceOk = "ext_authz_ok";
+};
+
+typedef ConstSingleton<ConstantValues> Constants;
+
+// TODO(htuch): We should have only one client per thread, but today we create one per filter stack.
+// This will require support for more than one outstanding request per client.
```

Comment:
Yeah, I see you're following the `RateLimit` pattern here. I'm not sure if it's the ideal one to copy, given that it has these limits. Probably fine for MVP, but it's not great that we're building on this.

Reply:
@hutch is there reference where there is a client per thread in existing code. I followed the `RateLimit` since it seemed closest to what `ExtAuthz` needed to do.
If it does not exist and if the patter of invoking external services is becoming common-which i suspect it is- should I open a issue/enhancement to track creating a gRPC client per thread?

Reply:
You can look at what https://github.com/envoyproxy/envoy/blob/master/source/common/access_log/grpc_access_log_impl.h does, I think that's close to what you want.

Reply:
thanks for the reference. let me take a look at it.

Reply:
@htuch I have a quick question on this. afaik the gRPC request is per thread and multiple `listener` filter stacks will use it. Will gRPC stack be able to take care of co-relating the `response` with the correct `request`? Or is this something that the `ThreadLocalCallers` class will have to handle with a `unordered_map` like [here]( (https://github.com/envoyproxy/envoy/blob/5f8b84179ffc1530476da4d8a250835fa1274a3c/source/common/access_log/grpc_access_log_impl.h#L101)
Even with that there is not a unique identifier in the gRPC request/response for ExtAuthz. So it's not clear to me how to map the response back to the calling/invoker filter.
Unless if gRPC already takes care of this under the covers?

Reply:
As mentioned above, I think it's different for authz, since you are unary RPC, so you can probably put this aside for now.